driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -1170,55 +1183,270 @@ int getNumPinnedToTransaction() {
+            boolean result = true;
+            if (state == State.NEW) {
+                getter = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncGetter"));
+                assertNotNull(getter).submit(() -> runAndLogUncaught(() -> workerRun(gets)));
+                opener = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncOpener"));
+                assertNotNull(opener).submit(() -> runAndLogUncaught(() -> workerRun(opens)));
+                timeouter = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncTimeouter"));
+                assertNotNull(timeouter).submit(() -> runAndLogUncaught(this::timeouterRun));
+                state = State.INITIALIZED;
+            } else if (state == State.CLOSED) {
+                result = false;
             }
-            return opener;
+            return result;
         }
 
+        /**
+         * {@linkplain Thread#interrupt() Interrupts} all workers and causes queued tasks to
+         * {@linkplain Task#failAsClosed() fail} asynchronously.
+         */
         @Override
+        @SuppressWarnings("try")
         public void close() {
+            lock.lock();
             try {
-                if (getter != null) {
-                    getter.shutdownNow();
+                if (state != State.CLOSED) {
+                    try (UncheckedAutoCloseable ignored3 = UncheckedAutoCloseable.create(getter);
+                         UncheckedAutoCloseable ignored2 = UncheckedAutoCloseable.create(opener);
+                         UncheckedAutoCloseable ignored1 = UncheckedAutoCloseable.create(timeouter)) {
+                        state = State.CLOSED;
+                        closedOrTaskAddedCondition.signalAll();
+                    } // at this point we interrupt worker threads
                 }
             } finally {
-                if (opener != null) {
-                    opener.shutdownNow();
+                lock.unlock();
+            }
+        }
+
+        private void workerRun(final BlockingQueue<Task> tasks) {
+            try {
+                while (state != State.CLOSED) {
+                    try {
+                        /* Take a task without holding the `lock`. As a result, we may complete the `task` concurrently with
+                         * other threads completing it. This is why `Task` allows to complete itself at most once. */
+                        Task task = tasks.take();
+                        if (task.timeout().expired()) {
+                            task.failAsTimedOut();
+                        } else {
+                            task.execute();
+                        }
+                    } catch (RuntimeException e) {
+                        LOGGER.error(null, e);
+                    }
                 }
+            } catch (InterruptedException closed) {
+                // fail the rest of the tasks and stop
             }
+            LOGGER.debug("Stopping");
```

Comment:
Nit: Logging seems inconsistent.  Suggest either remove the logging or improve the logging message and include it elsewhere if needed.